Create windows build distribution bundle in GHA

akrieger opened the pull request:
<!-- HOW TO USE: Under each "#### Heading" below, enter information relevant to your pull request.
Leave the headings unless they don't apply to your PR.

Please read carefully and don't delete the comments delimited by "< !--" and "-- >"
Once a pull request is submitted automatic stylistic and consistency checks will be performed on the PR's changes.
The results of these can be either seen under the "Files changed" section of a PR or in the check's details.

NOTE: Please grant permission for repository maintainers to edit your PR.  It is EXTREMELY common for PRs to be held up due to trivial changes being requested and the author being unavailable to make them. -->

#### Summary
None
<!-- This section should consist of exactly one line, edit the one above.
Category must be one of these: Features, Content, Interface, Mods, Balance, Bugfixes, Performance, Infrastructure, Build, I18N. Or replace the whole line with just the word None for no changelog entry.
For more on the meaning of each category, see:
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/CHANGELOG_GUIDELINES.md
If approved and merged, your summary will be added to the project changelog:
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/data/changelog.txt -->

#### Purpose of change
The MSX cross compiler was limiting our ability to use c++14 functionality like constexpr in some places. Long term it's a blocker to moving to c++17 because there isn't a newer prebuilt available and building a new version on the fly in every build is not palatable. As a result, it's beneficial to actually build our Windows builds on Windows machines from the VS projects themselves.
<!-- With a few sentences, describe your reasons for making this change.  If it relates to an existing issue, you can link it with a # followed by the GitHub issue number, like #1234.  If your pull request *fully* resolves an issue, include the word "Fix" or "Fixes" before the issue number, like: Fixes #1234.
If there is no related issue, explain here what issue, feature, or other concern you are addressing.  If this is a bugfix, include steps to reproduce the original bug, so your fix can be verified. -->

#### Describe the solution
Update `release.yml` to build the Windows builds on Windows workers using the same basic framework as the newly created Windows PR builds. Add a helper powershell script to assemble the distribution into a zip so we can make changes to the distribution without changing the actual build release file.
<!-- How does the feature work, or how does this fix a bug?  The easier you make your solution to understand, the faster it can get merged. -->

#### Describe alternatives you've considered
- Keep using MSX and try to update it (nah).
- Cross compile using clang, which would've necessitated setting up a full cross compile toolchain. Too much like my day job.
- Could've broken up `release.yml` into sub-actions. It might still be hygienic to do so in the future.
<!-- Explain any alternative solutions, different approaches, or possibilities you've considered using to solve the same problem. -->

#### Testing
https://github.com/akrieger/Cataclysm-DDA/releases/tag/cdda-experimental-2021-09-22-0546 has valid builds.
<!-- Describe what steps you took to test that this PR resolved the bug or added the feature, and what tests you performed to make sure it didn't cause any regressions.  Also include testing suggestions for reviewers and maintainers. -->

#### Additional context

Ping @jbytheway.

jbytheway commented on build-scripts/windist.ps1:
```diff
@@ -0,0 +1,13 @@
+$extras = "data", "doc", "gfx", "LICENSE.txt", "LICENSE-OFL-Terminus-Font.txt", "README.md", "VERSION.txt"
+ForEach ($extra in $extras) {
+	cp -r $extra bindist
+}
```
I would probably put the `Compress-Archive` step inside this script rather than in the workflow.

akrieger replied:
I had that at first, until I read about the upload-artifacts double zipping issue, so I wanted to leave option value open. But now that I know about release.yml I will look there instead and it’ll probably compress in this script again.

jbytheway commented:
Thanks for working on this.

Note for reviewers: this includes the changes from #51693.

akrieger commented on .github/workflows/msvc-full-features.yml:
```diff
@@ -95,7 +91,28 @@ jobs:
+    - name: Assemble windist
+      run: |
+          .\build-scripts\windist.ps1
+    # Although zipping first results in a double zip, we have to do this to
+    # avoid being ratelimited on upload because of the thousands of files
+    # in the dist bundle. This results in a zipped zip, but oh well.
```
I mostly wanted to get in something that worked at all, and be able to click and see that this is sane-ish, and try cleaning up the double zip situation later. I didn’t realize there’s a separate release.yml though, I’ll check that out.

akrieger commented on .github/workflows/msvc-full-features.yml:
```diff
@@ -95,7 +91,28 @@ jobs:
+    # in the dist bundle. This results in a zipped zip, but oh well.
+    - name: Compress windist
+      run: |
+          Compress-Archive -Path bindist/* -DestinationPath cataclysmdda-0.F
```
It does not. It’s documented that using the glob pattern sets the root as inside the folder. Also when I tested locally it was correct.

BrettDong commented on build-scripts/windist.ps1:
```diff
@@ -0,0 +1,13 @@
+mkdir bindist/lang
+cp -r lang/mo bindist/lang
+
+$extras = "data", "doc", "gfx", "LICENSE.txt", "LICENSE-OFL-Terminus-Font.txt", "README.md", "VERSION.txt"
```
I suggest excluding `data/mods/TEST_DATA` from release distribution.

akrieger replied:
That's a good idea, but the official distro also includes this data and for now at least I want to try to do as faithful of an adaptation as possible.

akrieger replied:
I'm initially trying to just match the existing distribution one for one, which does not remove the test data.

akrieger commented:
- No longer has someone else's changes misappropriated into it.
- Updated `release.yml` to build and ship cata.
- Fixed VS solution to properly embed manifest and thus have pretty icon on the final binary.
- Changed the VS solution to emit binaries of a consistent name for simplicity sake.

ZhilkinSerg commented on msvc-full-features/JsonFormatter-vcpkg-static.vcxproj:
```diff
@@ -125,4 +137,4 @@
   <ImportGroup Label="ExtensionTargets">
   </ImportGroup>
-</Project>
+</Project>
```
```suggestion
</Project>

```

BrettDong commented on msvc-full-features/Cataclysm-vcpkg-static.vcxproj:
```diff
@@ -117,6 +133,9 @@
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.targets" />
   <ImportGroup Label="ExtensionTargets" />
-</Project>
+</Project>
```
```suggestion
</Project>

```

akrieger commented:
- Restored mxe builds to original state, added msvc builds with `-msvc` added to artifact name.
- Addressed other feedback.

akrieger commented:
https://github.com/akrieger/Cataclysm-DDA/releases/tag/cdda-experimental-2021-09-30-0020 shows successful builds and I can run them.

jbytheway reviewed (approved):
This looks good to me.
